vchan/init.c

```diff
@@ -112,95 +106,181 @@ libvchan_t *libvchan_client_init(int domain, int port) {
+            goto err;
+        }
+        if (own_domid_num == ctrl->remote_domain) {
+            fprintf(stderr, "Loopback vchan connection not supported\n");
+            goto err;
+        }
+        ctrl->local_domain = own_domid_num;
+    }
 
-            v = snprintf(xs_path, sizeof(xs_path), "/local/domain/%d/data/vchan/%s/%d",
-                         domain, own_domid, port);
-            if ((unsigned)v >= sizeof(xs_path)) {
-                free(own_domid);
-                goto err_xs;
-            }
-            /* watch on this key as we might not have access to the whole directory */
-            v = snprintf(xs_path_watch, sizeof(xs_path_watch), "%.128s/event-channel", xs_path);
-            if ((unsigned)v >= sizeof(xs_path_watch)) {
-                free(own_domid);
-                goto err_xs;
-            }
+    /* construct xenstore path on first iteration and on every domid
+     * change detected (save+restore case) */
+    if (!ctrl->xs_path) {
+        int v;
 
-            if (!xs_watch(xs, xs_path_watch, xs_path_watch)) {
-                fprintf(stderr, "Cannot setup watch on %s\n", xs_path_watch);
-                free(own_domid);
-                goto err_xs;
-            }
+        v = snprintf(xs_path_base, sizeof(xs_path_base), "/local/domain/%d/data/vchan/%d/%d",
+                     ctrl->remote_domain, ctrl->local_domain, ctrl->port);
+        if ((unsigned)v >= sizeof(xs_path_base)) {
+            goto err;
         }
 
-        dummy = xs_read(xs, 0, xs_path_watch, &len);
-        if (dummy)
-            free(dummy);
-        else {
-            if (!libvchan__check_domain_alive(xc_handle, domain)) {
-                fprintf(stderr, "domain dead\n");
-                goto err_xs;
-            }
+        /* watch on this key as we might not have access to the whole directory */
+        v = asprintf(&ctrl->xs_path, "%.128s/event-channel", xs_path_base);
+        if (v < 0) {
+            goto err;
+        }
+
+        if (!xs_watch(ctrl->xs, ctrl->xs_path, ctrl->xs_path)) {
+            fprintf(stderr, "Cannot setup watch on %s\n", ctrl->xs_path);
+            goto err;
         }
     }
 
-    free(own_domid);
-    xs_close(xs);
+    tmp_str = xs_read(ctrl->xs, 0, ctrl->xs_path, &len);
+    if (tmp_str)
+        free(tmp_str);
+    else {
+        if (!libvchan__check_domain_alive(ctrl->xc_handle, ctrl->remote_domain)) {
+            fprintf(stderr, "domain dead\n");
+            goto err;
+        }
+        /* no xenstore entry (yet), wait more */
+        return 1;
+    }
 
-    ctrl = malloc(sizeof(*ctrl));
-    if (!ctrl)
-        goto err_xc;
+    if (!len)
+        /* empty entry, wait more */
+        return 1;
+
+    /* when got here, wait is over, attempt to connect */
+    xs_close(ctrl->xs);
+    ctrl->xs = NULL;
+
+    /* cut trailing /event-channel */
+    tmp_str = strrchr(ctrl->xs_path, '/');
+    assert(tmp_str);
+    *tmp_str = '\0';
```

Comment:
The `strrchr()` is redundant.  Just store the return value of `asprintf()` and subtract the correct number of bytes.

Reply:
The `asprintf()` might not be called at this iteration.